superset/connectors/sqla/models.py

```diff
@@ -2026,6 +2027,15 @@ class and any keys added via `ExtraCache`.
         if self.has_extra_cache_key_calls(query_obj):
             sqla_query = self.get_sqla_query(**query_obj)
             extra_cache_keys += sqla_query.extra_cache_keys
+
+        sqlalchemy_url = make_url_safe(self.database.sqlalchemy_uri_decrypted)
+
+        if effective_user := self.database.get_effective_user(sqlalchemy_url):
+            logger.debug("User impersonation is enabled for database '%s', adding "
+                         "current username to '%s' datasource's extra cache keys",
+                         self.database_name, self.datasource_name)
+            extra_cache_keys.append(effective_user)
```

Comment:
Instead of appending the effective user to `extra_cache_keys`, I'd almost prefer to have a dedicated key in the `cache_key_dict` for this that gets populated if `self.database.impersonate_user` is `True`. Also, to make it possible to customize this behavior per database type, id add functionality to `BaseEngineSpec` to make it possible to override the behavior. So something similar to this:

1. add a `get_impersonation_key(user: Optional[User]): Any` method to `BaseEngineSpec` that defaults to returning the username, but can be overridden where applicable. We may also want to add other objects to the signature, e.g. `Database`, but we could do that later.
2. In `QueryObject.cache_key()`, we would check if `self.datasource.database.impersonate_user` is truthy, and if so, we call `self.datasource.database.db_engine_spec.get_impersonation_key(user)` and add the value to `cache_dict["impersonation_key"]`

Thoughts?

Reply:
Hi Ville, I really appreciate your review of this PR!

I have no issue with your suggestion, let me make changes, test them out and get back to you.

Reply:
Hi @villebro, I made changes as per your suggestions, I ran into some issues so the impementation is not exactly as requested:

* I made `get_impersonation_key` expect a username as str rather than a User instance, I didn't want to add a trip to the DB to get the instance by username just to use the username eventually, I'm getting the username from the flask app context using the existing function `get_username`, if there is a better way then please lmk.
* I noticed that the `datasource` param in QueryObject is always null, it doesn't seem to be set when create an instance of this class, so I change the Factory class to explicity pass it, that way be able to use it to the db and impersonate_user flag. 
* As requested by nytai, there is now a feature flag to enable/disable this caching per user, it's off by default, not sure if the name is ok (naming stuff is hard 😄 ), happy to change the flag.


Can you take a look please? TIA